aten/src/ATen/cpu/vec256/vec256_float_neon.h

```diff
@@ -642,6 +642,853 @@ Vec256<float> inline fmadd(const Vec256<float>& a, const Vec256<float>& b, const
+              : [in_b] "w" (b.values.val[0]), [in_a] "w" (a.values.val[0])
+              :);
+          return vec;
+        }
+      case 4:
+        return Vec256<float>(b.values.val[0], a.values.val[1]);
+      case 5:
+        {
+          Vec256<float> vec;
+          static uint32x4_t mask_high = {0xFFFFFFFF, 0x0, 0x0, 0x0};
+          vec.values.val[0] = b.values.val[0];
+          vec.values.val[1] = (float32x4_t)mask_high;
+          __asm__ (
+              "vbsl.f32 %[res_mask], %[in_b], %[in_a]\n\t"
+              : [res_mask] "+w" (vec.values.val[1])
+              : [in_b] "w" (b.values.val[1]), [in_a] "w" (a.values.val[1])
+              :);
+          return vec;
+        }
+      case 6:
+        {
+          Vec256<float> vec;
+          static uint32x4_t mask_high = {0xFFFFFFFF, 0xFFFFFFFF, 0x0, 0x0};
+          vec.values.val[0] = b.values.val[0];
+          vec.values.val[1] = (float32x4_t)mask_high;
+          __asm__ (
+              "vbsl.f32 %[res_mask], %[in_b], %[in_a]\n\t"
+              : [res_mask] "+w" (vec.values.val[1])
+              : [in_b] "w" (b.values.val[1]), [in_a] "w" (a.values.val[1])
+              :);
+          return vec;
+        }
+      case 7:
+        {
+          Vec256<float> vec;
+          static uint32x4_t mask_high = {0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0x0};
+          vec.values.val[0] = b.values.val[0];
+          vec.values.val[1] = (float32x4_t)mask_high;
+          __asm__ (
+              "vbsl.f32 %[res_mask], %[in_b], %[in_a]\n\t"
+              : [res_mask] "+w" (vec.values.val[1])
+              : [in_b] "w" (b.values.val[1]), [in_a] "w" (a.values.val[1])
+              :);
+          return vec;
+        }
+    }
+    return b;
+  }
+  static Vec256<float> loadu(const void* ptr, int64_t count = size()) {
+    if (count == size()) {
+      volatile register float32x4_t low asm("q3");
+      volatile register float32x4_t high asm("q4");
+      __asm__ __volatile__ (
+          "vld1.32 {d6, d7, d8, d9}, [%[in_ptr]]\n\t"
+          : "=w" (low), "=w" (high)
+          : [in_ptr] "r" (ptr)
+          :);
+      return Vec256<float>(low, high);
+    }
+    else if (count == (size() >> 1)) {
+      volatile register float32x4_t low asm("q3");
+      float32x4_t high{0.f, 0.f, 0.f, 0.f};
+      Vec256<float> res;
+      __asm__  __volatile__ (
+          "vld1.32 {d6, d7}, [%[in_ptr]]\n\t"
+          : "=w" (low)
+          : [in_ptr] "r" (ptr)
+          :);
+      return Vec256<float>(low, high);
+    }
+    else {
+      __at_align32__ float tmp_values[size()];
+      for (auto i = 0; i < size(); ++i) {
+        tmp_values[i] = 0.0;
+      }
+      std::memcpy(
+          tmp_values,
+          reinterpret_cast<const float*>(ptr),
+          count * sizeof(float));
+      volatile register float32x4_t low asm("q3");
+      volatile register float32x4_t high asm("q4");
+      __asm__  __volatile__ (
+          "vld1.32 {d6, d7, d8, d9}, [%[in_ptr]]\n\t"
```

Comment:
You can specify the alignment as part of the instruction with `@align` now that you know the memory is aligned.

Reply:
Does it matter? And how?

Reply:
This is microarchitecture dependent but generally there is a performance benefit to aligned loads and stores.

> You can also specify an alignment for the pointer passed in Rn, using the optional : parameter, which often speeds up memory accesses.

Reference: https://community.arm.com/developer/ip-products/processors/b/processors-ip-blog/posts/coding-for-neon---part-1-load-and-stores

You can see that your particular use (i.e. VLDn {Dd, D(d+1), D(d+2), D(d+3)}, Rn) can accept three different, increasingly more strict, alignments in the table on [VLDn's reference page](https://developer.arm.com/documentation/dui0489/h/neon-and-vfp-programming/vldn-and-vstn--multiple-n-element-structures-?lang=en).

Reply:
My understanding was that, unaligned address cause perf penalty, especially vectorized code if your address happens to cross cacheline boundary. But if the address is already aligned then it does not matter.
A55 optimization manual seems to suggest that you benefit: https://static.docs.arm.com/epm128372/20/arm_cortex_a55_software_optimization_guide_v2.pdf (section 3.3) but only for LDRD etc. Does not say anything about vector loads.
A75's does not mention that.
I think this will be good to benchmark but it does not hurt to add that. So, I will add and benchmark later.